### PR TITLE
[Merged by Bors] - Refactor ECS to reduce the dependency on a 1-to-1 mapping between components and real rust types

### PR DESCRIPTION
**Pull request by bjorn3**
# Objective

There is currently a 1-to-1 mapping between components and real rust types. This means that it is impossible for multiple components to be represented by the same rust type or for a component to not have a rust type at all. This means that component types can't be defined in languages other than rust like necessary for scripting or sandboxed (wasm?) plugins.

## Solution

Refactor `ComponentDescriptor` and `Bundle` to remove `TypeInfo`. `Bundle` now uses `ComponentId` instead. `ComponentDescriptor` is now always created from a rust type instead of through the `TypeInfo` indirection. A future PR may make it possible to construct a `ComponentDescriptor` from it's fields without a rust type being involved.

### CODE REVIEW

**Comment by bjorn3** [comment]
cc @BoxyUwU re your relationships PR.

File: crates/bevy_ecs/src/component.rs [hunk 1]
```diff
@@ -154,14 +136,30 @@ pub struct ComponentDescriptor {
...
+
+    fn new_non_send<T: Any>(storage_type: StorageType) -> Self {
+        Self {
+            name: std::any::type_name::<T>().to_string(),
+            storage_type,
+            is_send_and_sync: false,
+            type_id: Some(TypeId::of::<T>()),
+            layout: Layout::new::<T>(),
+            drop: Self::drop_ptr::<T>,
         }
```
Comment (bjorn3) [inline crates/bevy_ecs/src/component.rs:163]: Making this public may help with https://github.com/bevyengine/bevy/issues/2486, but there are probably other places that need to be changed too for it to work.

File: crates/bevy_ecs/src/component.rs [hunk 2]
```diff
@@ -279,42 +269,58 @@ impl Components {
...
 
         ComponentId(*index)
     }
 
+    /// # Safety
+    ///
+    /// The [`ComponentDescriptor`] must match the [`TypeId`]
     #[inline]
-    pub(crate) fn get_or_insert_with(
+    pub(crate) unsafe fn get_or_insert_with(
```
Comment (bjorn3) [inline crates/bevy_ecs/src/component.rs:314]: This method was already de-facto unsafe. Nobody checked that the `TypeId` in the `TypeInfo` matched the `type_id` argument.

File: crates/bevy_ecs/src/component.rs [hunk 3]
```diff
@@ -279,42 +269,58 @@ impl Components {
...
+    /// # Safety
+    ///
+    /// The [`ComponentDescriptor`] must match the [`TypeId`]
     #[inline]
-    fn get_or_insert_resource_with(
+    unsafe fn get_or_insert_resource_with(
         &mut self,
         type_id: TypeId,
-        func: impl FnOnce() -> TypeInfo,
+        func: impl FnOnce() -> ComponentDescriptor,
```
Comment (BoxyUwU) [inline crates/bevy_ecs/src/component.rs:297]: do we know if perf is actually worse without this `func` param? can we just take a `ComponentDescriptor` always and drop the `type_id` arg?

Comment (bjorn3) [reply crates/bevy_ecs/src/component.rs:297]: `ComponentDescriptor` allocates a `String` for the type name. That is not something you want in the hot path.

Comment (BoxyUwU) [reply crates/bevy_ecs/src/component.rs:297]: ah right, the string... rip

Comment (BoxyUwU) [reply crates/bevy_ecs/src/component.rs:297]: type_name returns a `&'static str` though, seems like we could atleast use `Cow<&'static str>` here. Wouldn't be the best for custom names though :frown:

Comment (bjorn3) [reply crates/bevy_ecs/src/component.rs:297]: `Cow<'static, str>` would work. The name is not frequently used, so the extra branch shouldn't hurt. I will leave that for another PR though I think.

**Review by alice-i-cecile — APPROVED** [review]

File: crates/bevy_ecs/src/component.rs (hunk 3 above)
Comment (BoxyUwU) [reply crates/bevy_ecs/src/component.rs:297]: I think personally I'd rather drop the unsafe, use a Cow and then always take a `ComponentDescriptor`

Comment (bjorn3) [reply crates/bevy_ecs/src/component.rs:297]: `ComponentDescriptor` is 72 bytes with `String` and 80 bytes with Cow, which is relatively large. https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=93f494d4a982196b786f0f7d7f69c895

Comment (BoxyUwU) [reply crates/bevy_ecs/src/component.rs:297]: I dont know if that actually tells us anything about the performance here though?

**Review by cart — APPROVED** [review]
Looks good to me. Just some nits

File: crates/bevy_ecs/src/bundle.rs [hunk 4]
```diff
@@ -38,13 +38,13 @@ use std::{any::TypeId, collections::HashMap};
...
 /// _exact_ order that [Bundle::get_components] is called.
-/// [Bundle::from_components] must call `func` exactly once for each [TypeInfo] returned by
-/// [Bundle::type_info]
+/// [Bundle::from_components] must call `func` exactly once for each [ComponentId] returned by
+/// [Bundle::component_id]
 pub unsafe trait Bundle: Send + Sync + 'static {
-    /// Gets this [Bundle]'s components type info, in the order of this bundle's Components
-    fn type_info() -> Vec<TypeInfo>;
+    /// Gets this [Bundle]'s component ids, in the order of this bundle's Components
+    fn component_id(components: &mut Components) -> Vec<ComponentId>;
```
[the remark this answers is not in the document]
Comment (bjorn3) [reply crates/bevy_ecs/src/bundle.rs]: Renamed

**Comment by BoxyUwU** [comment]
bors retry

**Comment by cart** [comment]
nice that worked!

**Pull request closed without merging**